x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/PemUtils.java

```diff
@@ -0,0 +1,491 @@
+    private static RSAPrivateCrtKeySpec parseRsaDer(byte[] keyBytes) throws IOException {
+        DerParser parser = new DerParser(keyBytes);
+        DerParser.Asn1Object sequence = parser.readAsn1Object();
+        parser = sequence.getParser();
+        parser.readAsn1Object().getInteger(); // (version) We don't need it but must read to get to modulus
+        BigInteger modulus = parser.readAsn1Object().getInteger();
+        BigInteger publicExponent = parser.readAsn1Object().getInteger();
+        BigInteger privateExponent = parser.readAsn1Object().getInteger();
+        BigInteger prime1 = parser.readAsn1Object().getInteger();
+        BigInteger prime2 = parser.readAsn1Object().getInteger();
+        BigInteger exponent1 = parser.readAsn1Object().getInteger();
+        BigInteger exponent2 = parser.readAsn1Object().getInteger();
+        BigInteger coefficient = parser.readAsn1Object().getInteger();
+
+        return new RSAPrivateCrtKeySpec(modulus, publicExponent, privateExponent, prime1, prime2, exponent1, exponent2, coefficient);
+    }
+
+    /**
+     * Parses a DER encoded DSA key to a {@link DSAPrivateKeySpec} using a minimal {@link DerParser}
+     */
+    private static DSAPrivateKeySpec parseDsaDer(byte[] keyBytes) throws IOException {
+        DerParser parser = new DerParser(keyBytes);
+        DerParser.Asn1Object sequence = parser.readAsn1Object();
+        parser = sequence.getParser();
+        parser.readAsn1Object().getInteger(); // (version) We don't need it but must read to get to p
+        BigInteger p = parser.readAsn1Object().getInteger();
+        BigInteger q = parser.readAsn1Object().getInteger();
+        BigInteger g = parser.readAsn1Object().getInteger();
+        parser.readAsn1Object().getInteger(); // we don't need x
+        BigInteger x = parser.readAsn1Object().getInteger();
+
+        return new DSAPrivateKeySpec(x, p, q, g);
+    }
+
+    /**
+     * Parses a DER encoded private key and reads its algorithm identifier Object OID.
+     *
+     * @param keyBytes the private key raw bytes
+     * @return A string identifier for the key algorithm (RSA, DSA, or EC ) or null for any other algorithm
+     * @throws IOException
+     */
+    private static String getKeyAlgorithmIdentifier(byte[] keyBytes) throws IOException {
+        DerParser parser = new DerParser(keyBytes);
+        DerParser.Asn1Object sequence = parser.readAsn1Object();
+        parser = sequence.getParser();
+        parser.readAsn1Object().getInteger(); // version
+        DerParser.Asn1Object algSequence = parser.readAsn1Object();
+        parser = algSequence.getParser();
+        String oidString = parser.readAsn1Object().getOid();
+        switch (oidString) {
+            case "1.2.840.10040.4.1":
+                return "DSA";
+            case "1.2.840.113549.1.1.1":
+                return "RSA";
+            case "1.2.840.10045.2.1":
+                return "EC";
+            default:
+                return null;
```

Comment:
Both of the callers of this function translate `null` into an exception.   
If we throw it here then the exception can include the parsed OID string in order to provide context - either we got the ASN parsing wrong (and it would helpful to know what we parsed), or the key has an algorithm we don't support (and it would also helpful to know what the OID is)